components/file_system/src/rate_limiter.rs

```diff
@@ -1,57 +1,634 @@
+            }
+        }
+    }
+
+    pub fn fetch(&self, io_type: IOType, io_op: IOOp) -> usize {
+        match io_op {
+            IOOp::Read => self.read[io_type as usize].load(Ordering::Relaxed),
+            IOOp::Write => self.write[io_type as usize].load(Ordering::Relaxed),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn reset(&self) {
+        for i in self.read.iter() {
+            i.store(0, Ordering::Relaxed);
+        }
+        for i in self.write.iter() {
+            i.store(0, Ordering::Relaxed);
+        }
+    }
+}
+
+#[derive(Debug)]
+struct PerTypeIORateLimiter {
+    bytes_per_refill: AtomicUsize,
+    consumed: AtomicUsize,
+    refill_period: Duration,
+    last_refill_time: Mutex<Instant>,
+    condv: Condvar,
+    calibrator: Option<BytesCalibrator>,
+}
+
+#[inline]
+fn calculate_bytes_per_refill(bytes_per_sec: usize, refill_period: Duration) -> usize {
+    (bytes_per_sec as f64 * refill_period.as_secs_f64()) as usize
+}
+
+impl PerTypeIORateLimiter {
+    /// Create a new rate limiter. Rate limiting is disabled when `bytes_per_sec` is zero.
+    pub fn new(bytes_per_sec: usize, refill_period: Duration) -> PerTypeIORateLimiter {
+        PerTypeIORateLimiter {
+            bytes_per_refill: AtomicUsize::new(calculate_bytes_per_refill(
+                bytes_per_sec,
+                refill_period,
+            )),
+            consumed: AtomicUsize::new(0),
+            refill_period,
+            last_refill_time: Mutex::new(Instant::now_coarse()),
+            condv: Condvar::new(),
+            calibrator: None,
+        }
+    }
+
+    pub fn set_bytes_per_sec(&self, bytes_per_sec: usize) {
+        self.bytes_per_refill.store(
+            calculate_bytes_per_refill(bytes_per_sec, self.refill_period),
+            Ordering::Relaxed,
+        );
+    }
+
+    #[allow(dead_code)]
+    pub fn set_calibrator(&mut self, calibrator: BytesCalibrator) {
+        self.calibrator = Some(calibrator);
+    }
+
+    #[inline]
+    fn request_fast(&self, bytes_per_refill: usize, bytes: usize) -> Option<usize> {
+        if self.consumed.load(Ordering::Relaxed) < bytes_per_refill {
+            // Consumed bytes are allowed to be larger than the actual bytes
+            // through when quotas are drained.
+            let before = self.consumed.fetch_add(bytes, Ordering::Relaxed);
+            if before < bytes_per_refill {
+                return Some(std::cmp::min(bytes_per_refill - before, bytes));
+            }
+        }
+        None
+    }
+
+    #[inline]
+    fn refill_and_request(&self, bytes: usize) -> usize {
+        let token = std::cmp::min(self.bytes_per_refill.load(Ordering::Relaxed), bytes);
+        self.consumed.store(token, Ordering::Relaxed);
+        if let Some(calibrator) = &self.calibrator {
+            calibrator.reset();
+        }
+        self.condv.notify_all();
+        token
+    }
+
+    pub fn request(&self, bytes: usize, priority: IOPriority) -> usize {
```

Comment:
how about writing like `block_on(async_request)` to reuse the logic

Reply:
block_on a async function has a considerable overhead. The microbenchmark also shows a nearly 2x latency.